File: packages/i18nify-js/src/modules/geo/detectLocale.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { CountryCodeType } from '../types';
import { I18NIFY_DATA_SOURCE } from '../shared';
import { CountryMetaType } from './types';

type DetectLocaleOptions = {
  countryCode?: CountryCodeType;
  currency?: string;
  acceptLanguage?: string;
  browserLocale?: string;
};

// Parses RFC 7231 Accept-Language header, returns highest-priority locale tag.
const parseAcceptLanguage = (header: string): string | null => {
  const entries = header
    .split(',')
    .map((part) => {
      const [raw, q] = part.trim().split(';q=');
      const locale = raw.trim();
      return { locale, q: q !== undefined ? parseFloat(q) : 1.0 };
    })
    .filter(({ locale }) => locale.length > 0 && locale !== '*');

  if (entries.length === 0) return null;
  entries.sort((a, b) => b.q - a.q);
  return entries[0].locale;
};

const getLocaleFromCurrency = (
  currency: string,
  metadata: Record<CountryCodeType, CountryMetaType>,
): string | null => {
  const normalizedCurrency = currency.toUpperCase();
  const derivedCountryCode = normalizedCurrency.slice(0, 2) as CountryCodeType;

  const derivedCountry = metadata[derivedCountryCode];
  if (derivedCountry?.default_currency === normalizedCurrency) {
    return derivedCountry.default_locale;
  }

  const matchingLocales = Object.values(metadata)
    .filter((info) => info.default_currency === normalizedCurrency)
    .map((info) => info.default_locale)
    .filter(Boolean)
    .sort();

  return matchingLocales[0] ?? null;
};

/**
 * Detects a BCP 47 locale string from one or more signals.
 * Priority: countryCode → acceptLanguage → browserLocale → currency.
 *
 * @example
 * detectLocale({ countryCode: 'IN' })                        // → "en_IN"
 * detectLocale({ acceptLanguage: 'en-US,en;q=0.9' })        // → "en-US"
 * detectLocale({ browserLocale: 'de-DE' })                  // → "de-DE"
 * detectLocale({ currency: 'INR' })                         // → "en_IN"
 */
const detectLocale = (options: DetectLocaleOptions): Promise<string> => {
  const { countryCode, currency, acceptLanguage, browserLocale } = options;

  if (!countryCode && !currency && !acceptLanguage && !browserLocale) {
    return Promise.reject(
      new Error(
        'At least one detection signal must be provided: countryCode, currency, acceptLanguage, or browserLocale.',
      ),
    );
  }

  // Fast path: no metadata fetch needed
  if (!countryCode && !currency) {
    const locale =
      (acceptLanguage && parseAcceptLanguage(acceptLanguage)) ||
      browserLocale?.trim() ||
      null;
    return locale
      ? Promise.resolve(locale)
      : Promise.reject(
          new Error('Could not determine locale from the provided signals.'),
        );
  }

  return fetch(`${I18NIFY_DATA_SOURCE}/country/metadata/data.json`)
    .then((res) => res.json())
    .then(
      (res: {
        metadata_information: Record<CountryCodeType, CountryMetaType>;
      }) => {
        const meta = res.metadata_information;

        if (countryCode && meta[countryCode]?.default_locale) {
          return meta[countryCode].default_locale;
        }

        if (acceptLanguage) {
          const parsed = parseAcceptLanguage(acceptLanguage);
          if (parsed) return parsed;
        }

        if (browserLocale?.trim()) return browserLocale.trim();

        if (currency) {
          const locale = getLocaleFromCurrency(currency, meta);
          if (locale) return locale;
        }

        return Promise.reject(
          new Error('Could not determine locale from the provided signals.'),
        );
      },
    )
    .catch((err) => {
      throw new Error(
        `An error occurred while detecting locale. The error details are: ${err?.message || err}.`,
      );
    });
};

export default withErrorBoundary<typeof detectLocale>(detectLocale);

File: packages/i18nify-go/modules/geo/detect_locale.go
package geo

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// DetectLocaleOptions holds the available locale detection signals.
// At least one field must be non-empty. Signals are tried in priority order:
// CountryCode → AcceptLanguage → BrowserLocale → Currency.
type DetectLocaleOptions struct {
	CountryCode    string
	Currency       string
	AcceptLanguage string
	BrowserLocale  string
}

// DetectLocale resolves a BCP 47 locale string from one or more signals.
//
// Priority: CountryCode → AcceptLanguage → BrowserLocale → Currency.
//
// Example:
//
//	DetectLocale(DetectLocaleOptions{CountryCode: "IN"})                     // → "en_IN", nil
//	DetectLocale(DetectLocaleOptions{AcceptLanguage: "en-US,en;q=0.9"})     // → "en-US", nil
//	DetectLocale(DetectLocaleOptions{BrowserLocale: "de-DE"})               // → "de-DE", nil
//	DetectLocale(DetectLocaleOptions{Currency: "INR"})                      // → "en_IN", nil
func DetectLocale(opts DetectLocaleOptions) (string, error) {
	if opts.CountryCode == "" && opts.Currency == "" && opts.AcceptLanguage == "" && opts.BrowserLocale == "" {
		return "", fmt.Errorf("at least one detection signal must be provided: CountryCode, Currency, AcceptLanguage, or BrowserLocale")
	}

	if cachedCountryMetadata == nil {
		return "", fmt.Errorf("detectLocale: country metadata not loaded")
	}

	meta := cachedCountryMetadata.GetMetadataInformation()

	// Priority 1: CountryCode
	if opts.CountryCode != "" {
		cc := strings.ToUpper(strings.TrimSpace(opts.CountryCode))
		if info, ok := meta[cc]; ok && info.GetDefaultLocale() != "" {
			return info.GetDefaultLocale(), nil
		}
	}

	// Priority 2: AcceptLanguage header
	if opts.AcceptLanguage != "" {
		if locale := parseAcceptLanguageHeader(opts.AcceptLanguage); locale != "" {
			return locale, nil
		}
	}

	// Priority 3: BrowserLocale (already a locale string)
	if trimmed := strings.TrimSpace(opts.BrowserLocale); trimmed != "" {
		return trimmed, nil
	}

	// Priority 4: Currency reverse lookup
	if opts.Currency != "" {
		cu := strings.ToUpper(strings.TrimSpace(opts.Currency))

		// Prefer the ISO currency prefix when it maps back to a country using
		// the same default currency, e.g. USD -> US, INR -> IN.
		if len(cu) >= 2 {
			if info, ok := meta[cu[:2]]; ok && info != nil && strings.ToUpper(info.GetDefaultCurrency()) == cu && info.GetDefaultLocale() != "" {
				return info.GetDefaultLocale(), nil
			}
		}

		matchingLocales := make([]string, 0)
		for _, info := range meta {
			if info != nil && strings.ToUpper(info.GetDefaultCurrency()) == cu && info.GetDefaultLocale() != "" {
				matchingLocales = append(matchingLocales, info.GetDefaultLocale())
			}
		}

		if len(matchingLocales) > 0 {
			sort.Strings(matchingLocales)
			return matchingLocales[0], nil
		}
	}

	return "", fmt.Errorf("could not determine locale from the provided signals")
}

// parseAcceptLanguageHeader parses an RFC 7231 Accept-Language value and
// returns the highest-priority (q-value) locale tag, ignoring wildcards.
func parseAcceptLanguageHeader(header string) string {
	type entry struct {
		locale string
		q      float64
	}
	var entries []entry

	for _, part := range strings.Split(header, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		idx := strings.Index(part, ";q=")
		var locale string
		var q float64
		if idx == -1 {
			locale = strings.TrimSpace(part)
			q = 1.0
		} else {
			locale = strings.TrimSpace(part[:idx])
			parsed, err := strconv.ParseFloat(strings.TrimSpace(part[idx+3:]), 64)
			if err != nil {
				continue
			}
			q = parsed
		}
		if locale == "" || locale == "*" {
			continue
		}
		entries = append(entries, entry{locale: locale, q: q})
	}

	if len(entries) == 0 {
		return ""
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].q > entries[j].q })
	return entries[0].locale
}
